### DISCUSSION TITLE
kustomize-controller: Use cli-utils apply package

### DISCUSSION THREAD

**Discussion by seaneagan** [fluxcd/flux2#309] [Proposals] [answered]
The [cli-utils apply package](https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/apply) which is used by [kpt live apply](https://googlecontainertools.github.io/kpt/reference/live/apply/) has a bunch of useful functionality, some of which overlaps with kustomize-controller functionality, e.g. pruning. It uses an inventory object instead of labels to determine package membership. This has the advantage of not needing to modify the payload objects themselves. The inventory object id could be added to the Kustomization status. This could be a nice UX for determining what resources the last kustomization reconciliation produced. It could also allow for integration with e.g. `kpt live status` (beyond the existing kstatus health checks, as for example we can't include all status output in the CR status conditions).

**Comment by stefanprodan** [comment, marked as answer]
We label objects so that users can trace back what things on the cluster originated from which Kustomization. As for pruning, the cli utils doesn’t handle objects stuck in finlization, maybe for kpt which is a CLI is not a problem but for kustomize controller this means it can no longer reconcile new revisions, we have this problem with Flux v1 and it’s a major pain.

**Reply by seaneagan** [reply]
> We label objects so that users can trace back what things on the cluster originated from which Kustomization.

I think something similar could be accomplished using an inventory object whose namespace/name is:

1. based on the Kustomization namespace/name
2. included in a Kustomization status object field

Except that if pruning is disabled, there could be unidentified objects leftover. Could also just keep the labels, but use the inventory objects for pruning, if we can figure out the below issue.

> As for pruning, the cli utils doesn’t handle objects stuck in finlization, maybe for kpt which is a CLI is not a problem but for kustomize controller this means it can no longer reconcile new revisions, we have this problem with Flux v1 and it’s a major pain.

The pruning has some other niceties like calculating an appropriate ordering of the deletions (reverse of applications). I wonder if handling objects stuck in finalization could be added to cli-utils or could be tacked on in kustomize-controller after calling cli-utils. I will try to file an issue there to see.

**Reply by stefanprodan** [reply]
> I think something similar could be accomplished using an inventory object

Users shoudn't have to deal with an inventory object, right now they can do it with just kubectl:

```
$ kubectl get deployments \
-l=kustomize.toolkit.fluxcd.io/name=test,kustomize.toolkit.fluxcd.io/namespace=gotk-system \
--all-namespaces 
```

> The pruning has some other niceties like calculating an appropriate ordering of the deletions

Kustomize controller does its own calculation such that CRs can be finalized before deleting the CRDs.

What issues have you encountered in kustomize-controller's GC that would be solved with cli-utils?

**Reply by seaneagan** [reply]
> Users shoudn't have to deal with an inventory object, right now they can do it with just kubectl:
> 
> ```
> $ kubectl get deployments \
> -l=kustomize.toolkit.fluxcd.io/name=test,kustomize.toolkit.fluxcd.io/namespace=gotk-system \
> --all-namespaces 
> ```

I agree the labels are definitely useful for that. The inventory object in cli-utils is specifically about the pruning use case.
 
> What issues have you encountered in kustomize-controller's GC that would be solved with cli-utils?

None at all, just wanted to look into the possibility of aligning on a common library with similar functionality, but sounds like it may not satisfy the needs of kustomize-controller. Thanks for the responses!

**Reply by stefanprodan** [reply]
I'm for reusing kustomize libraries as much as we can, but GC is a super sensitive area that I think we should approach from a controller perspective rather than a CLI/interactive way. Looking ahead, we will probably need to extend our GC to treat CAPI objects differently, using some library that we don't control will make it super hard to deal with edge cases.

**Reply by seaneagan** [reply]
@stefanprodan #324 is another use case for the inventory object. What do you think about using cli-utils (for the inventory object) without using its pruning functionality?

**Reply by stefanprodan** [reply]
Why would we need the the inventory object for that? The Kustomize GC snapshot has all the data for selecting the objects by kind/label and feed them to the kstatus.
